src/main/java/org/spongepowered/common/effect/SpongeViewer.java

```diff
@@ -0,0 +1,148 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.effect;
+
+import net.kyori.adventure.sound.SoundStop;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.title.Title;
+import net.kyori.adventure.title.TitlePart;
+import net.minecraft.network.protocol.game.ClientboundBlockDestructionPacket;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.api.Engine;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.effect.Viewer;
+import org.spongepowered.api.effect.particle.ParticleEffect;
+import org.spongepowered.api.effect.sound.music.MusicDisc;
+import org.spongepowered.api.world.WorldType;
+import org.spongepowered.common.bridge.effect.ViewerBridge;
+import org.spongepowered.common.effect.util.ViewerPacketUtil;
+
+import java.util.Objects;
+
+public interface SpongeViewer extends Viewer {
+
+    // Viewer
+
+    @Override
+    default void sendWorldType(final WorldType worldType) {
+        ((ViewerBridge) this).bridge$sendSpongePacketToViewer(ViewerPacketUtil.changeEnvironment(worldType));
+    }
+
+    @Override
+    default void spawnParticles(final ParticleEffect particleEffect, final double x, final double y, final double z) {
+        ((ViewerBridge) this).bridge$sendToViewer(ViewerPacketUtil.spawnParticles(particleEffect, x, y, z));
+    }
+
+    @Override
+    default void playMusicDisc(final int x, final int y, final int z, final MusicDisc musicDisc) {
+    }
+
+    @Override
+    default void stopMusicDisc(final int x, final int y, final int z) {
+        ((ViewerBridge) this).bridge$sendToViewer(ViewerPacketUtil.stopMusicDisc(x, y, z));
+    }
+
+    @Override
+    default void sendBlockChange(final int x, final int y, final int z, final BlockState state) {
+        ((ViewerBridge) this).bridge$sendToViewer(ViewerPacketUtil.blockUpdate(x, y, z, state));
+    }
+
+    @Override
+    default void resetBlockChange(final int x, final int y, final int z) {
+    }
+
+    @Override
+    default void sendBlockProgress(final int x, final int y, final int z, final double progress) {
+        final Engine engine;
```

Comment:
Your current implementation may pick the `Server` engine when calling the method on a `ClientPlayer`.
IMO Engine should be determined by the viewer type. Eg:
`ServerPlayer` -> `Server` engine
`ClientPlayer` -> `Client` engine
You could add a `engine()` method in this interface, and implement it in the subclasses.

Reply:
Ended up with direct implementation. Luckily we already have `World#engine()`